extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -123,50 +115,106 @@ export class Bind {
+      this.evaluator_ = new BindEvaluator(evaluatees);
+
+      // Trigger verify-only digest in development.
+      if (getMode().development || this.digestQueuedAfterScan_) {
+        this.digest_(/* opt_verifyOnly */ !this.digestQueuedAfterScan_);
+      }
+    });
+  }
+
+  /**
    * Scans `body` for attributes that conform to bind syntax and returns
    * a tuple containing bound elements and binding data for the evaluator.
    * @param {!Element} body
-   * @return {{
-   *   boundElements: !Array<BoundElementDef>,
-   *   evaluatees: !Array<./bind-evaluator.EvaluateeDef>,
-   * }}
+   * @return {
+   *   !Promise<{
+   *     boundElements: !Array<BoundElementDef>,
+   *     evaluatees: !Array<./bind-evaluator.EvaluateeDef>
+   *   }>
+   * }
    * @private
    */
   scanForBindings_(body) {
-    // TODO(choumx): Chunk if taking too long in a single frame.
-
     /** @type {!Array<BoundElementDef>} */
     const boundElements = [];
     /** @type {!Array<./bind-evaluator.EvaluateeDef>} */
     const evaluatees = [];
 
+    // TODO(choumx): Use TreeWalker if this is too slow.
     const elements = body.getElementsByTagName('*');
-    for (let i = 0; i < elements.length; i++) {
-      const element = elements[i];
-      const attributes = element.attributes;
-
-      const bindings = [];
-      for (let j = 0; j < attributes.length; j++) {
-        const binding = this.bindingForAttribute_(attributes[j], element);
-        if (binding) {
-          bindings.push(binding);
-          evaluatees.push({
-            tagName: element.tagName,
-            property: binding.property,
-            expressionString: binding.expressionString,
-          });
+
+    // Helper function for scanning a slice of elements.
+    const scanFromTo = (start, end) => {
+      for (let i = start, el = elements[i]; i < end && el; i++) {
+        const attrs = el.attributes;
+        const bindings = [];
+        for (let j = 0; attrs[j]; j++) {
+          const binding = this.bindingForAttribute_(attrs[j], el);
+          if (binding) {
+            bindings.push(binding);
+            evaluatees.push({
+              tagName: el.tagName,
+              property: binding.property,
+              expressionString: binding.expressionString,
+            });
+          }
+        }
+        if (bindings.length > 0) {
+          boundElements.push({element: el, bindings});
         }
       }
-      if (bindings.length > 0) {
-        boundElements.push({element, bindings});
-      }
-    }
-    return {boundElements, evaluatees};
+    };
+
+    // Current scan position in `elements` array.
+    let position = 0;
+
+    return new Promise(resolve => {
+      const chunktion = idleDeadline => {
+        // If `requestIdleCallback` is available, scan elements until
+        // idle time runs out.
+        if (idleDeadline && !idleDeadline.didTimeout) {
+          while (idleDeadline.timeRemaining() > 1 && elements[position]) {
```

Comment:
Oh, it's not as bad if extracting the inner loop only; done. 

Mutating input params is generally bad for readability but I think it's ok in this case.